File: src/python/clean_data.py
"""
clean_data.py
Loads Bloomberg25MAR.xlsx, pivots to long format, removes NaNs,
computes daily returns and lags, and saves to data/processed/panel_long.csv.

Output columns (one row per ticker x trading day):
  date, ticker, px_open, px_close, px_high, px_low, mkt_cap, total_equity,
  debt_to_equity, volume, twitter_sent, twitter_count, news_sent,
  rsi_30, ma_50, twitter_neg_count, return, lag1, lag2, lag3, lag5, lag7
"""

import pandas as pd
import numpy as np
from pathlib import Path

# ---------------------------------------------------------------------------
# Paths
# ---------------------------------------------------------------------------
ROOT      = Path(__file__).resolve().parents[2]
RAW_FILE  = ROOT / "data" / "raw" / "Bloomberg25MAR.xlsx"
OUT_FILE  = ROOT / "data" / "processed" / "panel_long.csv"

OUT_FILE.parent.mkdir(parents=True, exist_ok=True)


# ===========================================================================
# BLOCK 1 — LOAD & PIVOT
# ===========================================================================

raw = pd.read_excel(RAW_FILE, header=None)

# Row 3 (0-indexed): ticker names, one per 14-column block
# Row 5 (0-indexed): Bloomberg field codes (PX_OPEN, PX_OFFICIAL_CLOSE, ...)
h1 = raw.iloc[3].astype(str).str.strip().replace("nan", np.nan)   # tickers
h3 = raw.iloc[5].astype(str).str.strip().replace("nan", np.nan)   # field codes

# Fill ticker name across its 14 sub-columns
h1 = pd.Series(h1).ffill()

# Keep col 0 (dates) + all cols that have a real field code in row 5
keep_mask = [True] + [pd.notna(h3.iloc[i]) and h3.iloc[i] != "Dates"
                      for i in range(1, len(h3))]
keep_idx  = [i for i, k in enumerate(keep_mask) if k]

raw = raw.iloc[:, keep_idx].reset_index(drop=True)
h1  = h1.iloc[keep_idx].reset_index(drop=True)
h3  = h3.iloc[keep_idx].reset_index(drop=True)

# Data starts at row 6 (rows 0-5 are metadata/headers)
dat = raw.iloc[6:].reset_index(drop=True).copy()

# Build column names as "TICKER__FIELD"
col_names = ["date"] + [
    f"{ticker}__{field}"
    for ticker, field in zip(h1.iloc[1:], h3.iloc[1:])
]
seen = {}
unique_names = []
for name in col_names:
    if name in seen:
        seen[name] += 1
        unique_names.append(f"{name}__dup__{seen[name]}")
    else:
        seen[name] = 0
        unique_names.append(name)
dat.columns = unique_names

# Convert Excel serial date numbers to proper dates
dat["date"] = pd.to_datetime(
    pd.to_numeric(dat["date"], errors="coerce"),
    unit="D",
    origin="1899-12-30"
).dt.normalize()

# Pivot wide → long, then spread fields back to columns
long = dat.melt(id_vars="date", var_name="ticker__field", value_name="value")
long[["ticker", "field"]] = long["ticker__field"].str.split("__", n=1, expand=True)
long = long.drop(columns="ticker__field")

long = long.pivot_table(
    index=["date", "ticker"],
    columns="field",
    values="value",
    aggfunc="first"
).reset_index()
long.columns.name = None

# Rename Bloomberg field codes to snake_case
long = long.rename(columns={
    "PX_OPEN":                      "px_open",
    "PX_OFFICIAL_CLOSE":            "px_close",
    "PX_HIGH":                      "px_high",
    "PX_LOW":                       "px_low",
    "CUR_MKT_CAP":                  "mkt_cap",
    "TOTAL_EQUITY":                 "total_equity",
    "TOT_DEBT_TO_TOT_EQY":         "debt_to_equity",
    "PX_VOLUME":                    "volume",
    "TWITTER_SENTIMENT_DAILY_AVG":  "twitter_sent",
    "TWITTER_PUBLICATION_COUNT":    "twitter_count",
    "NEWS_SENTIMENT_DAILY_AVG":     "news_sent",
    "RSI_30D":                      "rsi_30",
    "MOV_AVG_50D":                  "ma_50",
    "TWITTER_NEG_SENTIMENT_COUNT":  "twitter_neg_count",
})

long = long.sort_values(["ticker", "date"]).reset_index(drop=True)


# ===========================================================================
# BLOCK 2 — REMOVE NaNs
# ===========================================================================

BLOOMBERG_NA = ["#N/A N/A", "#N/A", "#N/A Field Not Applicable"]

numeric_cols = [
    "px_open", "px_close", "px_high", "px_low", "mkt_cap", "total_equity",
    "debt_to_equity", "volume", "twitter_sent", "twitter_count", "news_sent",
    "rsi_30", "ma_50", "twitter_neg_count"
]

for col in numeric_cols:
    if col in long.columns:
        long[col] = long[col].replace(BLOOMBERG_NA, np.nan)
        long[col] = pd.to_numeric(long[col], errors="coerce")

# Drop non-trading days (weekends/holidays are all-NA for prices)
long = long.dropna(subset=["px_open", "px_close"]).reset_index(drop=True)


# ===========================================================================
# BLOCK 3 — DAILY RETURN + LAGS
# ===========================================================================

long["return"] = long["px_close"] - long["px_open"]

for n in [1, 2, 3, 5, 7]:
    long[f"lag{n}"] = long.groupby("ticker")["return"].shift(n)


# ===========================================================================
# SAVE
# ===========================================================================

long.to_csv(OUT_FILE, index=False)
print(f"Saved {len(long):,} rows x {long.shape[1]} cols → {OUT_FILE}")
print(long.dtypes)
print(long.head(10).to_string())

File: src/r/clean_data.R
# clean_data.R
# R equivalent of src/python/clean_data.py
#
# Loads Bloomberg25MAR.xlsx, pivots to long format, removes NaNs,
# computes daily returns and lags, and saves to data/processed/panel_long.csv.
#
# Output columns (one row per ticker x trading day):
#   date, ticker, px_open, px_close, px_high, px_low, mkt_cap, total_equity,
#   debt_to_equity, volume, twitter_sent, twitter_count, news_sent,
#   rsi_30, ma_50, twitter_neg_count, return, lag1, lag2, lag3, lag5, lag7

library(readxl)
library(dplyr)
library(tidyr)
library(stringr)
library(purrr)

# ---------------------------------------------------------------------------
# Paths
# ---------------------------------------------------------------------------
ROOT     <- normalizePath(file.path(dirname(rstudioapi::getSourceEditorContext()$path), "../.."))
RAW_FILE <- file.path(ROOT, "data", "raw", "Bloomberg25MAR.xlsx")
OUT_FILE <- file.path(ROOT, "data", "processed", "panel_long.csv")

dir.create(dirname(OUT_FILE), showWarnings = FALSE, recursive = TRUE)


# ===========================================================================
# BLOCK 1 — LOAD & PIVOT
# ===========================================================================

raw <- read_excel(RAW_FILE, sheet = 1, col_names = FALSE)

# Row 4 (1-indexed) = row 3 (0-indexed): ticker names, one per 14-column block
# Row 6 (1-indexed) = row 5 (0-indexed): Bloomberg field codes
h1 <- raw[4, ] %>% unlist(use.names = FALSE) %>% as.character()   # tickers
h3 <- raw[6, ] %>% unlist(use.names = FALSE) %>% as.character()   # field codes

h1 <- na_if(str_trim(h1), "NA")
h3 <- na_if(str_trim(h3), "NA")
h3 <- na_if(h3, "Dates")

# Fill ticker name across its 14 sub-columns
h1 <- tidyr::fill(tibble(x = h1), x, .direction = "down")$x

# Keep col 1 (dates) + all cols that have a real field code in row 6
keep_cols <- c(1, which(!is.na(h3) & seq_along(h3) != 1))

raw <- raw[, keep_cols]
h1  <- h1[keep_cols]
h3  <- h3[keep_cols]

# Data starts at row 7 (rows 1-6 are metadata/headers)
dat <- raw[-c(1, 2, 3, 4, 5, 6), ]

names(dat)[1] <- "date"
names(dat)    <- c("date", map2_chr(h1[-1], h3[-1], ~ paste0(.x, "__", .y)))
names(dat)    <- make.unique(names(dat), sep = "__dup__")

# Convert Excel serial date numbers to proper dates
dat <- dat %>%
  mutate(date = as.Date(as.numeric(date), origin = "1899-12-30"))

# Pivot wide → long, then spread fields back to columns
long <- dat %>%
  pivot_longer(
    cols      = -date,
    names_to  = c("ticker", "field"),
    names_sep = "__",
    values_to = "value"
  ) %>%
  pivot_wider(names_from = field, values_from = value)

# Rename Bloomberg field codes to snake_case
long <- long %>%
  rename(
    px_open         = PX_OPEN,
    px_close        = PX_OFFICIAL_CLOSE,
    px_high         = PX_HIGH,
    px_low          = PX_LOW,
    mkt_cap         = CUR_MKT_CAP,
    total_equity    = TOTAL_EQUITY,
    debt_to_equity  = TOT_DEBT_TO_TOT_EQY,
    volume          = PX_VOLUME,
    twitter_sent    = TWITTER_SENTIMENT_DAILY_AVG,
    twitter_count   = TWITTER_PUBLICATION_COUNT,
    news_sent       = NEWS_SENTIMENT_DAILY_AVG,
    rsi_30          = RSI_30D,
    ma_50           = MOV_AVG_50D,
    twitter_neg_count = TWITTER_NEG_SENTIMENT_COUNT
  ) %>%
  arrange(ticker, date)


# ===========================================================================
# BLOCK 2 — REMOVE NaNs
# ===========================================================================

BLOOMBERG_NA <- c("#N/A N/A", "#N/A", "#N/A Field Not Applicable")

numeric_cols <- c(
  "px_open", "px_close", "px_high", "px_low", "mkt_cap", "total_equity",
  "debt_to_equity", "volume", "twitter_sent", "twitter_count", "news_sent",
  "rsi_30", "ma_50", "twitter_neg_count"
)

long <- long %>%
  mutate(across(all_of(numeric_cols), ~ na_if(as.character(.x), BLOOMBERG_NA[1]))) %>%
  mutate(across(all_of(numeric_cols), ~ na_if(.x, BLOOMBERG_NA[2]))) %>%
  mutate(across(all_of(numeric_cols), ~ na_if(.x, BLOOMBERG_NA[3]))) %>%
  mutate(across(all_of(numeric_cols), ~ as.numeric(.x)))

# Drop non-trading days (weekends/holidays are all-NA for prices)
long <- long %>%
  filter(!is.na(px_open), !is.na(px_close))


# ===========================================================================
# BLOCK 3 — DAILY RETURN + LAGS
# ===========================================================================

long <- long %>%
  mutate(return = px_close - px_open)

long <- long %>%
  arrange(ticker, date) %>%
  group_by(ticker) %>%
  mutate(
    lag1 = lag(return, 1),
    lag2 = lag(return, 2),
    lag3 = lag(return, 3),
    lag5 = lag(return, 5),
    lag7 = lag(return, 7)
  ) %>%
  ungroup()


# ===========================================================================
# SAVE
# ===========================================================================

write.csv(long, OUT_FILE, row.names = FALSE)
cat(sprintf("Saved %d rows x %d cols -> %s\n", nrow(long), ncol(long), OUT_FILE))
print(str(long))
print(head(long, 10))


